### PR TITLE
Closes #2686 Fix AZCardWidget deprecation warnings and broken default (empty) URL.

### LINKED ISSUE #2686: Card link field deprecation fix was incomplete
## Problem/Motivation
The fix merged for #2639 was incomplete.  We fixed the deprecation warning(s) getting generated by the card FieldFormatter plugin but not the same deprecation warning(s) being generated by the FieldWidget plugin.

When editing an existing card with a card link title value but no card link URL value, a deprecation warning is still generated on PHP 8.1.

### Describe the bug
Partial stack trace:
```
Deprecated function: ltrim(): Passing null to parameter #1 ($string) of type string is deprecated in Drupal\Core\Path\PathValidator->getUrl() (line 96 of /code/web/core/lib/Drupal/Core/Path/PathValidator.php)
#0 /code/web/core/includes/bootstrap.inc(158): _drupal_error_handler_real(8192, 'ltrim(): Passin...', '/code/web/core/...', 96) 
#1 [internal function]: _drupal_error_handler(8192, 'ltrim(): Passin...', '/code/web/core/...', 96) 
#2 /code/web/core/lib/Drupal/Core/Path/PathValidator.php(96): ltrim(NULL, '/') 
#3 /code/web/core/lib/Drupal/Core/Path/PathValidator.php(82): Drupal\Core\Path\PathValidator->getUrl(NULL, true) 
#4 /code/web/profiles/custom/az_quickstart/modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php(170): Drupal\Core\Path\PathValidator->getUrlIfValid(NULL) 
#5 /code/web/core/lib/Drupal/Core/Field/WidgetBase.php(349): Drupal\az_card\Plugin\Field\FieldWidget\AZCardWidget->formElement(Object(Drupal\Core\Field\FieldItemList), 0, Array, Array, Object(Drupal\Core\Form\FormState)) 
#6 /code/web/core/lib/Drupal/Core/Field/WidgetBase.php(216): Drupal\Core\Field\WidgetBase->formSingleElement(Object(Drupal\Core\Field\FieldItemList), 0, Array, Array, Object(Drupal\Core\Form\FormState)) 
#7 /code/web/core/lib/Drupal/Core/Field/WidgetBase.php(111): Drupal\Core\Field\WidgetBase->formMultipleElements(Object(Drupal\Core\Field\FieldItemList), Array, Object(Drupal\Core\Form\FormState))
#8 /code/web/profiles/custom/az_quickstart/modules/custom/az_card/src/Plugin/Field/FieldWidget/AZCardWidget.php(93): Drupal\Core\Field\WidgetBase->form(Object(Drupal\Core\Field\FieldItemList), Array, Object(Drupal\Core\Form\FormState), NULL) 
#9 /code/web/core/lib/Drupal/Core/Entity/Entity/EntityFormDisplay.php(183): Drupal\az_card\Plugin\Field\FieldWidget\AZCardWidget->form(Object(Drupal\Core\Field\FieldItemList), Array, Object(Drupal\Core\Form\FormState)) 
#10 /code/web/modules/contrib/paragraphs/src/Plugin/Field/FieldWidget/ParagraphsWidget.php(781): Drupal\Core\Entity\Entity\EntityFormDisplay->buildForm(Object(Drupal\paragraphs\Entity\Paragraph), Array, Object(Drupal\Core\Form\FormState)) 
#11 /code/web/core/lib/Drupal/Core/Field/WidgetBase.php(349): Drupal\paragraphs\Plugin\Field\FieldWidget\ParagraphsWidget->formElement(Object(Drupal\entity_reference_revisions\EntityReferenceRevisionsFieldItemList), 0, Array, Array, Object(Drupal\Core\Form\FormState)) 
#12 /code/web/modules/contrib/paragraphs/src/Plugin/Field/FieldWidget/ParagraphsWidget.php(1116): Drupal\Core\Field\WidgetBase->formSingleElement(Object(Drupal\entity_reference_revisions\EntityReferenceRevisionsFieldItemList), 0, Array, Array, Object(Drupal\Core\Form\FormState)) 
#13 /code/web/core/lib/Drupal/Core/Field/WidgetBase.php(111): Drupal\paragraphs\Plugin\Field\FieldWidget\ParagraphsWidget->formMultipleElements(Object(Drupal\entity_reference_revisions\EntityReferenceRevisionsFieldItemList), Array, Object(Drupal\Core\Form\FormState)) 
#14 /code/web/modules/contrib/paragraphs/src/Plugin/Field/FieldWidget/ParagraphsWidget.php(1238): Drupal\Core\Field\WidgetBase->form(Object(Drupal\entity_reference_revisions\EntityReferenceRevisionsFieldItemList), Array, Object(Drupal\Core\Form\FormState), NULL) 
#15 /code/web/core/lib/Drupal/Core/Entity/Entity/EntityFormDisplay.php(183): Drupal\paragraphs\Plugin\Field\FieldWidget\ParagraphsWidget->form(Object(Drupal\entity_reference_revisions\EntityReferenceRevisionsFieldItemList), Array, Object(Drupal\Core\Form\FormState)) 
#16 /code/web/core/lib/Drupal/Core/Entity/ContentEntityForm.php(121): Drupal\Core\Entity\Entity\EntityFormDisplay->buildForm(Object(Drupal\node\Entity\Node), Array, Object(Drupal\Core\Form\FormState)) 
#17 /code/web/core/modules/node/src/NodeForm.php(127): Drupal\Core\Entity\ContentEntityForm->form(Array, Object(Drupal\Core\Form\FormState)) 
#18 /code/web/core/lib/Drupal/Core/Entity/EntityForm.php(106): Drupal\node\NodeForm->form(Array, Object(Drupal\Core\Form\FormState)) 
#19 [internal function]: Drupal\Core\Entity\EntityForm->buildForm(Array, Object(Drupal\Core\Form\FormState)) 
#20 /code/web/core/lib/Drupal/Core/Form/FormBuilder.php(536): call_user_func_array(Array, Array)
```

### To Reproduce

1. Create flexible page
2. Add cards paragraph
3. Add card with a link title but no link URL
4. Save page
5. View page
6. Edit page (no need to save again)
7. Find deprecation warning in logs

## Proposed resolution
Use NULL coalescing operator on card link URI value passed to `ltrim()` (via `Drupal\Core\Path\PathValidator->getUrlIfValid()`).
### Expected behavior

No deprecation warnings.

## Additional context
https://www.php.net/manual/en/migration81.deprecated.php

### PR DESCRIPTION
**Pull request by joeparsons**
## Description
- PR #2641 fixed a deprecation warning generated by `AZCardFieldFormatter` but it turns out `AZCardWidget` also generates the same deprecation warning.  This should fix that deprecation warning.
- I think should also fix an issue with empty card links getting rendered as links to the front page.

## Related issues

Closes #2686 

## How to test

1. Create flexible page
2. Add cards paragraph
3. Add card with a link title but no link URL
4. Save page
5. View page
6. Verify that link is rendered with an empty `href` attribute (will behave like a link to the current page at least in Chrome)
7. Edit page (no need to save again)
8. Ensure no deprecation warning in logs

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [x] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Review by bberndt-uaz — DISMISSED** [review]
Tested in Probo as directed in the PR description and did not see any deprecation errors in the log.

**Review by ejsamboy — APPROVED** [review]

**Comment by joeparsons** [comment]
Changed the fallback value passed to the `pathValidator::getUrlIfValid()` method when URL field is empty from empty string to `<none>` since empty string (and previously `NULL`) seemed to be returning the front page URL.

Using `<none>` causes the card links without URL values to get rendered with an empty `href` attribute which browsers seem to render as a link to the current page.

I was hoping to be able to render a link to and empty URL fragment `#` but that doesn't seem possible using the Drupal URL / Link classes.

File: modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php [hunk 1]
```diff
@@ -122,11 +123,11 @@ public function viewElements(FieldItemListInterface $items, $langcode) {
       // Link.
       $link_render_array = [];
       if ($item->link_title || $item->link_uri) {
-        $link_url = $this->pathValidator->getUrlIfValid($item->link_uri ?? '');
+        $link_url = $this->pathValidator->getUrlIfValid($item->link_uri ?? '<none>');
```
Comment (joeparsons) [inline modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php:126]: Fallback to `<none>` for value passed to validation method (if `link_uri` value is empty)

File: modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php [hunk 2]
```diff
@@ -122,11 +123,11 @@ public function viewElements(FieldItemListInterface $items, $langcode) {
...
         $link_render_array = [
           '#type' => 'link',
           '#title' => $item->link_title ?? '',
-          '#url' => $link_url ? $link_url : '#',
+          '#url' => $link_url ?: Url::fromRoute('<none>'),
```
Comment (joeparsons) [inline modules/custom/az_card/src/Plugin/Field/FieldFormatter/AZCardDefaultFormatter.php:130]: Fallback to the URL for the `<none>` route if the validation (above) failed.

**Review by bberndt-uaz — APPROVED** [review]

**Review by ejsamboy — APPROVED** [review]

**Review by trackleft — APPROVED** [review]

**Pull request merged**